patches/tModLoader/Terraria/GameContent/Liquid/LiquidEdgeRenderer.cs

```diff
@@ -366,4 +386,57 @@ public static unsafe void CollectEdgeData(LiquidRenderer.LiquidCache* pCache, Ti
 			pCache->Type = (byte)liquidType;
 		}
 	}
+
+	private static void AddEdgePoint(ushort tileX, ushort tileY, bool hasSpecialMask)
+	{
+		// Masks always break the span because they need to be rendered separately.
+		if (hasSpecialMask) {
+			if (currentSpan.HasValue) {
+				edgeSpans.Add(currentSpan.Value);
+				currentSpan = null;
+			}
+
+			maskPoints.Add(new Point16(tileX, tileY));
+			return;
+		}
+
+		// Begin building the span if there is none.
+		if (!currentSpan.HasValue) {
+			currentSpan = new EdgeSpan {
+				X = tileX,
+				YStart = tileY,
+				Height = 1,
+			};
```

Comment:
These 4 lines are duplicated in 3 locations, there should be a way to write this method so that this is the common fallthrough path. Probably by making `if (currentSpan is {} span && span.X == tileX && span.YStart + span.Height == tileY) {currentSpan = span with { Height = span.Height + 1}; return;}` the first condition.